firebase-firestore/src/main/java/com/google/firebase/firestore/util/AsyncQueue.java

```diff
@@ -446,11 +447,30 @@ public void runDelayedTasksUntil(TimerId lastTimerId) throws InterruptedExceptio
 
   /**
    * Shuts down the AsyncQueue and releases resources after which no progress will ever be made
-   * again.
+   * again. Since we can't use the AsyncQueue for the callbacks, we'll instead use the main thread.
    */
-  public void shutdown() {
-    // Will cause the executor to de-reference all threads, the best we can do
-    executor.setCorePoolSize(0);
+  public Task<Void> shutdown() {
+    executor.shutdown();
+    return Tasks.call(
+        () -> {
+          if (!executor.awaitTermination(100, TimeUnit.MILLISECONDS)) {
```

Comment:
> Given network RTT on mobile networks can easily be 250+ ms I think this is way too tight a bound. This could reasonably be seconds.

Whoops; yeah, I meant to go back and fix that. Changed to 60s.

> Secondarily, are you sure we even need to shutdownNow? Have we observed things getting wedged?

No. But I'm following the example from java's docs, i.e. https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html

(Though not following closely enough for the actual executor case, i.e. the thread.interrupt thing.)